compiler/qsc_frontend/src/resolve.rs

```diff
@@ -131,7 +135,91 @@ impl Scope {
     }
 }
 
-struct GlobalScope {
+pub(super) type ScopeId = usize;
+
+#[derive(Debug, Clone, Default)]
+pub struct Locals {
+    // order is ascending by span (outermost -> innermost)
+    scopes: Vec<Scope>,
+}
+
+impl Locals {
+    pub(super) fn get_scopes<'a>(
+        &'a self,
+        scope_chain: &'a [ScopeId],
+    ) -> impl Iterator<Item = &Scope> + 'a {
+        // reverse to go from innermost -> outermost
+        scope_chain.iter().rev().map(|id| {
+            self.scopes
+                .get(*id)
+                .unwrap_or_else(|| panic!("scope with id {id:?} should exist"))
+        })
+    }
+
+    pub(super) fn push_scope(&mut self, s: Scope) -> ScopeId {
+        let id = self.scopes.len();
+        self.scopes.insert(id, s);
+        id
+    }
+
+    pub(super) fn get_scope_mut(&mut self, id: ScopeId) -> &mut Scope {
+        self.scopes
+            .get_mut(id)
+            .unwrap_or_else(|| panic!("scope with id {id:?} should exist"))
+    }
+
+    #[must_use]
+    pub fn get_all_at_offset(&self, offset: u32) -> Vec<Local> {
+        let mut vars = true;
+        let mut all_locals = Vec::new();
+        self.for_each_scope_at_offset(offset, |scope| {
+            // inner to outer
+            all_locals.extend(get_scope_locals(scope, offset, vars));
+
+            if scope.kind == ScopeKind::Callable {
+                // Since local callables are not closures, hide local variables in parent scopes.
+                vars = false;
+            }
+        });
+
+        // deduping by name will effectively make locals in a child scope
+        // shadow the locals in its parent scopes
+        all_locals.dedup_by(|a, b| a.name == b.name);
```

Comment:
I think you just posted a comment about this, but how can we be sure the right duplicate will get dropped here based on the cursor position?

Reply:
Ah, yeah it's related to the other bug I commented on. When you have a variable shadowed in the _same_ scope:

```qsharp
let var = 3;
let var = "foo";
```

By the time we get to this code that you commented on, the first declaration doesn't even exist in the `Scope` anymore. This is because `vars` is kept as a hashmap, and the second declaration overwrites the first one, with its own `valid_at` offset.

So, deduping within the same scope isn't a concern here, because there is no dupe. Which is itself a bug. ([this one](https://github.com/microsoft/qsharp/pull/880#discussion_r1419537650)) . I'm completely stumped on how to fix it in this PR and would like to open a separate GH issue for it.